src/Solnet.Programs/NameServiceProgram.cs

```diff
@@ -0,0 +1,285 @@
+        /// <returns>The transaction instruction.</returns>
+        /// <exception cref="Exception">Thrown when it was not possible to derive a program address for the account.</exception>
+        public static TransactionInstruction CreateNameRegistry(
+            PublicKey name, Account payer, PublicKey nameOwner, ulong lamports, int space, Account nameClass = null, 
+            Account parentNameOwner = null, PublicKey parentName = null)
+        {
+            byte[] hashedName = ComputeHashedName(name.Key);
+            Console.WriteLine("Hashed Name: " + Convert.ToHexString(hashedName).ToLowerInvariant());
+            PublicKey nameAccountKey = DeriveNameAccountKey(hashedName, nameClass?.PublicKey, parentName);
+            Console.WriteLine("Name Account Key: " + nameAccountKey.Key);
+            if (nameAccountKey == null) throw new Exception("could not derive an address for the name account");
+            return CreateNameRegistryInstruction(
+                nameAccountKey, nameOwner, payer, hashedName, lamports, space, nameClass, parentNameOwner, parentName);
+        }
+        
+        /// <summary>
+        /// Gets the hash for the given value with the attached hash prefix.
+        /// </summary>
+        /// <param name="name">The name to hash attach the prefix and hash.</param>
+        /// <returns>The hash as bytes.</returns>
+        public static byte[] ComputeHashedName(string name)
+        {
+            string prefixedName = HashPrefix + name;
+            Console.WriteLine(prefixedName);
+            byte[] fullNameBytes = Encoding.UTF8.GetBytes(prefixedName);
+            using SHA256Managed sha = new();
+            return sha.ComputeHash(fullNameBytes, 0, fullNameBytes.Length);
+        }
+        
+        /// <summary>
+        /// Get's the program derived address for the name.
+        /// </summary>
+        /// <param name="hashedName">The hash of the name with the name service hash prefix.</param>
+        /// <param name="nameClass">The account of the name class.</param>
+        /// <param name="parentName">The public key of the parent name.</param>
+        /// <returns>The program derived address for the name.</returns>
+        public static PublicKey DeriveNameAccountKey(byte[] hashedName, PublicKey nameClass = null, PublicKey parentName = null)
+        {
+            byte[] nameClassKey = new byte[32];
+            byte[] parentNameKeyBytes = new byte[32];
+
+            if (nameClass != null) nameClassKey = nameClass.KeyBytes;
+            if (parentName != null) parentNameKeyBytes = parentName.KeyBytes;
+
+            try
+            {
+                (byte[] nameAccountPublicKey, _) = AddressExtensions.FindProgramAddress(
+                    new List<byte[]> {hashedName, nameClassKey, parentNameKeyBytes}, ProgramIdKey.KeyBytes);
+                return new PublicKey(nameAccountPublicKey);
+            }
+            catch (Exception)

```

Comment:
what could cause an exception here? 
It should be very well documented as there is no other output/logs of why/how this might fail